apps/browser-extension/src/background/background.ts

```diff
@@ -1,124 +1,262 @@
+      linkUrl: undefined,
+      pageUrl: tab?.url,
+    });
+
+    // now try to open the popup
+    chrome.action.openPopup();
+  } else {
+    console.warn(`Received unknown command: ${command}`);
+  }
+}
+
+chrome.commands.onCommand.addListener(handleCommand);
+
+/**
+ * Set the badge text and color based on the provided information.
+ * @param text The text to display on the badge.
+ * @param isExisted Whether the badge should indicate existence.
+ * @param tabId The ID of the tab to update.
+ */
+export async function setBadge(
+  text: string | number,
+  isExisted: boolean,
+  tabId?: number,
+) {
+  return await Promise.all([
+    chrome.action.setBadgeText({ tabId, text: `${text}` }),
+    chrome.action.setBadgeBackgroundColor({
+      tabId,
+      color: isExisted ? "#4CAF50" : "#F44336",
+    }),
+  ]);
+}
+
+/**
+ * Get the count of bookmarks for a given tab URL.
+ * @param tabUrl The URL of the tab to check.
+ */
+export async function getTabCount(tabUrl: string) {
+  const api = await getApiClient();
+  const data = await api.bookmarks.searchBookmarks.query({
+    text: "url:" + tabUrl,
+  });
+  if (!data) {
+    return { count: 0, isExisted: false };
+  }
+  const bookmarks = data.bookmarks || [];
+  const isExisted = bookmarks.some(
+    (b) => b.content.type === BookmarkTypes.LINK && tabUrl === b.content.url,
+  );
+  return {
+    count: bookmarks.length,
+    isExisted,
+  };
+}
+
+/**
+ * Check and update the badge icon for a given tab ID.
+ * @param tabId The ID of the tab to update.
+ */
+async function checkAndUpdateIcon(tabId: number) {
+  const tabInfo = await chrome.tabs.get(tabId);
+  const pluginSettings = await getPluginSettings();
+  if (
+    !pluginSettings.showCountBadge ||
+    !tabInfo.url ||
+    tabInfo.status !== "complete"
+  ) {
+    return;
+  }
+  console.log("Tab activated", tabId, tabInfo);
+
+  try {
+    const cachedInfo = await getBadgeStatusSWR(tabInfo.url);
+    if (cachedInfo) {
+      const { count: cachedBadgeCount, isExisted: cachedIsExisted } =
+        cachedInfo;
+      await setBadge(cachedBadgeCount, cachedIsExisted, tabId);
+      return;
+    }
+    const { count, isExisted } = await getTabCount(tabInfo.url);
+    await setBadge(count, isExisted, tabId);
+    await setBadgeStatusSWR(tabInfo.url, count, isExisted);
+  } catch (error) {
+    console.error("Archive check failed:", error);
+    await setBadge("!", false, tabId);
+  }
+}
+
+chrome.tabs.onActivated.addListener(async (activeInfo) => {
+  await checkAndUpdateIcon(activeInfo.tabId);
+});
+
+chrome.tabs.onUpdated.addListener(async (changeInfo) => {
+  await checkAndUpdateIcon(changeInfo);
+});
+
+// Listen for REFRESH_BADGE messages from popup and update badge accordingly
+chrome.runtime.onMessage.addListener(async (msg) => {
+  if (msg && msg.type) {
+    if (
+      msg.currentTab &&
+      (msg.type === MessageType.BOOKMARK_CREATED_REFRESH_BADGE ||
+        msg.type === MessageType.BOOKMARK_DELETED_REFRESH_BADGE)
```

Comment:
If we're handling them the same, do we need two different message types? Why not just `BOOKMARK_REFRESH_BADGE`?